src/main/java/at/ac/tuwien/kr/alpha/grounder/IndexedInstanceStorage.java

```diff
@@ -168,6 +169,41 @@ public void removeInstance(Instance instance) {
 		return matchingInstances == null ? new ArrayList<>() : matchingInstances;
 	}
 
+
+	private int getFirstGroundTermPosition(Atom atom) {
+		for (int i = 0; i < atom.getTerms().size(); i++) {
+			Term testTerm = atom.getTerms().get(i);
+			if (testTerm.isGround()) {
+				return i;
+			}
+		}
+		return -1;
+	}
+
+	public Collection<Instance> getInstancesFromPartiallyGroundAtom(Atom substitute) {
```

Comment:
This signature combined with the returned expressions in lines 189 and 191 is a little strange. Since `getInstancesMatchingAtPosition` returns `List<Instance>` and `new ArrayList<>(getAllInstances()` will be a `List<Instance>` I would expect the method to return this type. If you prefer to return `Collection<Instance>` please justify this with a comment. Similar reasoning applies to `getInstancesMatching` below.